docs/sql/sql.md

```diff
@@ -0,0 +1,498 @@
+
+> NOTE: Arithmetic expressions will be supported in the future.
+
+## Query and DML
+
+### Statements
+
+- [SELECT](#select)
+- [INSERT](#insert)
+- [DELETE](#delete)
+- [UPDATE](#update)
+- [SET](#set)
+- [SHOW](#show)
+
+### SELECT
+
+```ebnf
+SELECT [ DISTINCT ] projection FROM relation [ [ WHERE predicate ] [ ORDER BY order ] [ LIMIT limit ] ]
+```
+
+#### SELECT Clause
+
+```ebnf
+projection
+    = '*'
+    | table '.' '*'
+    | projExpr { ',' projExpr }
+    | aggrExpr { ',' aggrExpr }
+    ;
+
+projExpr
+    = column [ [ AS ] alias ]
+    ;
+
+aggrExpr
+    = COUNT '(' STAR ')' [AS] alias
+    | COUNT '(' DISTINCT column ')' [AS] alias
+    | SUM   '(' column ')' [AS] alias
+    ;
+```
+
+The `SELECT` clause determines the columns that are returned.
+In particular it supports both column and wildcard projections.
+It also supports `sum` and `count` aggregation.
+
+If `DISTINCT` is specified, only unique rows are returned.
+However each projected column must be of a type that allows comparison.
+
+Examples:
+```sql
+-- Select the items in my inventory
+SELECT * FROM Inventory;
+
+-- Select the names of the items in my inventory
+SELECT item_name FROM Inventory
+
+-- How many items are in my inventory?
+SELECT COUNT(*) as num_items FROM Inventory
+```
+
+#### FROM Clause
+
+```ebnf
+FROM table [ [AS] alias ] { [INNER] JOIN table [ [AS] alias ] ON predicate }
+```
+
+> NOTE: The query language supports joins among an arbitrary number of tables.
+> It does not impose any index restrictions like the subscription language.
+
+Examples:
+```sql
+-- Which orders include currently stocked items?
+SELECT order.*
+FROM Orders order
+JOIN Inventory inv
+ON order.item_id = inv.item_id
+```
+
+#### WHERE Clause
+
+```ebnf
+predicate
+    = expr
+    | predicate AND predicate
+    | predicate OR  predicate
+    ;
+
+expr
+    = literal
+    | column
+    | expr op expr
+    ;
+
+op
+    = '='
+    | '<'
+    | '>'
+    | '<' '='
+    | '>' '='
+    | '!' '='
+    | '<' '>'
+    ;
+
+literal
+    = INTEGER
+    | FLOAT
+    | STRING
+    | HEX
+    | TRUE
+    | FALSE
+    ;
+```
+
+The query language supports both conjunctions and disjunctions.
+It also supports `!=` comparisons.
+
+> NOTE: Arithmetic expressions will be supported in the future.
+
+Examples:
+```sql
+SELECT * FROM Inventory WHERE item_id = 1;
+```
+
+#### ORDER BY Clause
+
+```ebnf
+ORDER BY column [ ASC | DESC ] { ',' column [ ASC | DESC ] }
+```
+
+`ORDER BY` sorts the result set by one or more of the projected columns.
+Ascending is the default behavior if not qualified.
+Each column involved must be of a type that allows comparison.
+Comparison is not supported for sum, product, or array types.
+
+Examples:
+```sql
+-- Sort the items in the inventory by order of increasing price
+SELECT * FROM Inventory ORDER BY price;
+
+-- Sort the items in the inventory by order of decreasing price
+SELECT * FROM Inventory ORDER BY price DESC
+
+#### LIMIT Clause
+
+```ebnf
+LIMIT INTEGER
+```
```

Comment:
Should we define an unsigned `INTEGER` type?

Reply:
We could.